File: FW 4.0/ESystem/_Extensions/ICollectionExt.vb
Option Explicit On
Option Strict On

Imports System.Runtime.CompilerServices

Namespace Extensions

  Public Module ICollectionExt

    ''' <summary>
    ''' Returns index of last item in collection
    ''' </summary>
    <Extension()> _
    Public Function LastIndex(ByVal arr As ICollection) As Integer
      Return arr.Count - 1
    End Function

    ''' <summary>
    ''' Returns items joined together by separator.
    ''' </summary>
    ''' <param name="separator"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    <Extension()> _
    Public Function ToString(ByVal arr As ICollection, ByVal separator As String) As String
      Dim ret As New Text.StringBuilder()

      For i As Integer = 0 To arr.Count - 1
        If i > 0 Then
          ret.Append(separator)
        End If
        ret.Append(arr(i))
      Next

      Return ret.ToString()
    End Function

    ''' <summary> 
    ''' Cuts list into two parts by index. 
    ''' </summary> 
    ''' <param name="index">Index where to split.</param> 
    ''' <param name="excludeSplittingItem">If True, indexed item will not appear in any created lists. 
    ''' If False, indexed item will be the first in the second list.</param> 
    ''' <returns></returns> 
    <Extension()> _
    Public Function SplitByIndex(Of T)(ByVal coll As ICollection(Of T), ByVal index As Integer, ByVal excludeSplittingItem As Boolean) As List(Of T)()
      Dim ret As List(Of T)() = New List(Of T)(1) {}

      ret(0) = coll.EGetRange(0, index)

      If excludeSplittingItem Then
        ret(1) = coll.EGetRange(index + 1)
      Else
        ret(1) = coll.EGetRange(index)
      End If

      Return ret
    End Function

    ''' <summary> 
    ''' Returns items in range from collection, from index to the end of source list. 
    ''' </summary> 
    ''' <param name="startIndex">Start index, inclusive.</param> 
    ''' <returns></returns> 
    <Extension()> _
    Public Function EGetRange(Of T)(ByVal coll As ICollection(Of T), ByVal startIndex As Integer) As List(Of T)
      Dim ret As List(Of T) = coll.EGetRange(startIndex, coll.Count - startIndex)

      Return ret
    End Function

    ''' <summary> 
    ''' Returns items in range from collection. 
    ''' </summary> 
    ''' <param name="startIndex">Start index, inclusive.</param> 
    ''' <param name="count">Count.</param> 
    ''' <returns></returns> 
    <Extension()> _
    Public Function EGetRange(Of T)(ByVal coll As ICollection(Of T), ByVal startIndex As Integer, ByVal count As Integer) As List(Of T)
      Dim ret As New List(Of T)

      For index As Integer = startIndex To startIndex + count - 1
        ret.Add(coll(index))
      Next

      Return ret
    End Function

    ''' <summary> 
    ''' Returns random item from collection. 
    ''' </summary> 
    ''' <returns></returns> 
    <Extension()> _
    Public Function GetRandom(Of T)(ByVal coll As ICollection(Of T), ByVal rnd As Random) As T
      If coll.Count = 0 Then
        Throw New Exception("Unable to get random item from empty collection.")
      Else
        Return coll(rnd.Next(0, coll.Count))
      End If
    End Function

    ''' <summary> 
    ''' Returns random item from collection. 
    ''' </summary> 
    ''' <returns></returns> 
    <Extension()> _
    Public Function GetRandom(Of T)(ByVal coll As ICollection(Of T)) As T
      Return GetRandom(coll, New Random())
    End Function
    ''' <summary> 
    ''' Returns random item from collection. 
    ''' </summary> 
    ''' <returns></returns> 
    <Extension()> _
    Public Function GetRandom(Of T)(ByVal coll As ICollection(Of T), ByVal seed As Integer) As T
      Return GetRandom(coll, New Random(seed))
    End Function

    <Extension()> _
    Public Function FindAll(Of T)(ByVal coll As IEnumerable(Of T), ByVal predicate As System.Func(Of T, Boolean)) As List(Of T)
      Dim ret As New List(Of T)

      For Each fItem As T In coll
        If predicate(fItem) Then ret.Add(fItem)
      Next

      Return ret

    End Function

    ''' <summary>
    ''' Removes item accepted by predicate from current collection and 
    ''' returns them in new collection.
    ''' </summary>
    ''' <param name="predicate">Predicate to match item to be removed.</param>
    ''' <returns>Removed items in new list.</returns>
    ''' <remarks></remarks>
    <Extension()> _
    Public Function Cut(Of T)(ByVal coll As ICollection(Of T), ByVal predicate As System.Func(Of T, Boolean)) As List(Of T)
      Dim ret As New List(Of T)

      ret = FindAll(coll, predicate)
      RemoveRange(coll, ret)

      Return (ret)
    End Function

    ''' <summary>
    ''' Removes item accepted by predicate from current collection into the parameter.
    ''' </summary>
    ''' <param name="predicate">Predicate to match item to be removed.</param>
    ''' <returns>Removed items in new list.</returns>
    ''' <remarks></remarks>
    <Extension()> _
    Public Function CutTo(Of T)(ByVal coll As ICollection(Of T), ByVal targetCollection As ICollection(Of T), ByVal predicate As System.Func(Of T, Boolean)) As List(Of T)
      Dim ret As New List(Of T)

      ret = FindAll(coll, predicate)
      RemoveRange(coll, ret)

      For Each fItem As T In ret
        targetCollection.Add(fItem)
      Next

      Return (ret)
    End Function

    <Extension()> _
    Public Sub RemoveRange(Of T)(ByVal coll As ICollection(Of T), ByVal items As IEnumerable(Of T))

      For Each fItem As T In items
        If coll.Contains(fItem) Then
          coll.Remove(fItem)
        End If
      Next

    End Sub

    <Extension()> _
Public Sub RemoveRange(Of T)(ByVal coll As ICollection(Of T), ByVal predicate As System.Func(Of T, Boolean))

      For index As Integer = coll.Count - 1 To 0 Step -1
        If (predicate(coll(index))) Then
          coll.Remove(coll(index))
        End If
      Next

    End Sub

    <Extension()> _
Public Sub LeaveRange(Of T)(ByVal coll As ICollection(Of T), ByVal predicate As System.Func(Of T, Boolean))

      For index As Integer = coll.Count - 1 To 0 Step -1
        If (Not predicate(coll(index))) Then
          coll.Remove(coll(index))
        End If
      Next

    End Sub


  End Module

End Namespace

File: _Previous/FW 3.5/ESystem/_Extensions/ICollectionExt.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ESystem.Extensions
{
  public static class ICollectionExt
  {

    /// <summary>
    /// Returns index of last item in collection
    /// </summary>
    public static int LastIndex(this ICollection arr)
    {
      return arr.Count - 1;
    }    

    /// <summary>
    /// Returns all items in enumeration fulfilling predicate conditions in new list.
    /// </summary>
    /// <typeparam name="T"></typeparam>
    /// <param name="coll"></param>
    /// <param name="predicate"></param>
    /// <returns></returns>
    public static List<T> FindAll<T>(this IEnumerable<T> coll, System.Func<T, bool> predicate)
    {
      List<T> ret = new List<T>();

      foreach (T fItem in coll)
      {
        if (predicate(fItem))
          ret.Add(fItem);
      }

      return ret;

    }

    /// <summary>
    /// Removes item fulfiled by predicate from current collection and 
    /// returns them in new collection.
    /// </summary>
    /// <param name="predicate">Predicate to match item to be removed.</param>
    /// <returns>Removed items in new list.</returns>
    /// <remarks></remarks>
    public static List<T> Cut<T>(this ICollection<T> coll, System.Func<T, bool> predicate)
    {
      List<T> ret = new List<T>();

      ret = FindAll(coll, predicate);
      RemoveRange(coll, ret);

      return (ret);
    }

    /// <summary>
    /// Removes item fulfiled by predicate from current collection into the parameter.
    /// </summary>
    /// <param name="predicate">Predicate to match item to be removed.</param>
    /// <returns>Removed items in new list.</returns>
    /// <remarks></remarks>
    public static List<T> CutTo<T>(this ICollection<T> coll,
      ICollection<T> targetCollection, System.Func<T, bool> predicate)
    {
      List<T> ret = new List<T>();

      ret = FindAll(coll, predicate);
      RemoveRange(coll, ret);

      foreach (T fItem in ret)
      {
        targetCollection.Add(fItem);
      }

      return (ret);
    }

    /// <summary>
    /// Removes items from collection, which are in other enumeration.
    /// </summary>
    /// <typeparam name="T"></typeparam>
    /// <param name="coll"></param>
    /// <param name="items"></param>
    public static void RemoveRange<T>(this ICollection<T> coll, IEnumerable<T> items)
    {
      foreach (T fItem in items)
      {
        if (coll.Contains(fItem))
        {
          coll.Remove(fItem);
        }
      }
    }

    /// <summary>
    /// Removes from the collection items fulfiling the predicate.
    /// </summary>
    /// <typeparam name="T"></typeparam>
    /// <param name="coll"></param>
    /// <param name="predicate"></param>
    public static void RemoveRange<T>(this ICollection<T> coll, System.Func<T, bool> predicate)
    {
      List<T> remList = new List<T>();

      foreach (var fItem in coll)
      {
        if (predicate(fItem))
          remList.Add(fItem);
      } // foreach (var fItem in coll)

      foreach (var fItem in remList)
      {
        coll.Remove(fItem);
      } // foreach (var fItem in remList)

    }

    /// <summary>
    /// Leaves in collection only items fulfilling the predicate.
    /// </summary>
    /// <typeparam name="T"></typeparam>
    /// <param name="coll"></param>
    /// <param name="predicate"></param>
    public static void LeaveRange<T>(this ICollection<T> coll, System.Func<T, bool> predicate)
    {
      coll.RemoveRange(i => (!predicate(i)));
    }

    /// <summary> 
    /// Adds new item into collection, if item is not already inserted. 
    /// Return true if item inserted, false if not. 
    /// </summary> 
    /// <param name="item"></param> 
    public static bool AddDistinct<T>(this ICollection<T> coll, T item)
    {
      if (!coll.Contains(item))
      {
        coll.Add(item);
        return true;
      }
      else
      {
        return false;
      }
    }

    /// <summary> 
    /// Adds new item into collection, if item is not already inserted. 
    /// </summary> 
    /// <param name="items"></param> 
    public static void AddDistinct<T>(this ICollection<T> coll, IEnumerable<T> items)
    {
      items.EForEach((T i) => AddDistinct(coll, i));
    }

  }
}
